src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java

```diff
@@ -267,6 +289,17 @@ public MqttClient publish(String topic, Buffer payload, MqttQoS qosLevel, boolea
 
     io.netty.handler.codec.mqtt.MqttMessage publish = MqttMessageFactory.newMessage(fixedHeader, variableHeader, buf);
 
+    switch (qosLevel) {
+      case AT_LEAST_ONCE:
+        qos1outbound.put(variableHeader.messageId(), publish);
+        countInflightQueue++;
+        break;
+      case EXACTLY_ONCE:
+        qos2outbound.put(variableHeader.messageId(), publish);
+        countInflightQueue++;
+        break;
+    }
+
     this.write(publish);
 
     if (publishSentHandler != null) {
```

Comment:
Sorry you are right. I forgot about that. For now I would not delete the publishSentHandler; it's useful to the user to know which is the message-id assigned to the publish so that he can do some check when the publishCompleteHandler is called.